File: Difficulty: Medium/Sort a linked list of 0s, 1s and 2s/sort-a-linked-list-of-0s-1s-and-2s.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class Node {
    int data;
    Node next;

    Node(int x) {
        data = x;
        next = null;
    }
}


// } Driver Code Ends

/* class Node
{
    int data;
    Node next;
    Node(int data)
    {
        this.data = data;
        next = null;
    }
}*/
class Solution {
    static Node segregate(Node head) {
        // code here
         int zeros=0, ones=0, twos=0;
        Node head1=head;
        while(head1!=null){
            if(head1.data==0){
                zeros++;
            }
            if(head1.data==1){
                ones++;
            }
            if(head1.data==2){
                twos++;
            }
            head1=head1.next;
        }
        head1=head;
        while(head1 !=null){
            if(zeros != 0){
                head1.data=0;
                zeros--;
            }
           else if(ones != 0){
                head1.data=1;
                ones--;
            }
           else if(twos != 0){
                head1.data=2;
                twos--;
            }
            head1=head1.next;
        }
        return head;
    }
}



//{ Driver Code Starts.

class GFG {
    public static void printList(Node node) {
        while (node != null) {
            System.out.print(node.data + " ");
            node = node.next;
        }
        System.out.println();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim());

        while (t-- > 0) {
            List<Integer> arr = new ArrayList<>();
            String input = br.readLine().trim();
            StringTokenizer st = new StringTokenizer(input);
            while (st.hasMoreTokens()) {
                arr.add(Integer.parseInt(st.nextToken()));
            }

            Node head = null;
            if (!arr.isEmpty()) {
                head = new Node(arr.get(0));
                Node tail = head;
                for (int i = 1; i < arr.size(); i++) {
                    tail.next = new Node(arr.get(i));
                    tail = tail.next;
                }
            }
            head = new Solution().segregate(head);
            printList(head);
            System.out.println("~");
        }
    }
}
// } Driver Code Ends

File: Difficulty: Medium/Sort a linked list of 0s, 1s and 2s/sort-a-linked-list-of-0s-1s-and-2s.cpp
/* Node is defined as
  class Node {
  public:
    int data;
    Node* next;

    Node(int x) {
        data = x;
        next = nullptr;
    }
};
*/
class Solution {
  public:
    Node* segregate(Node* head) {
        // code here
        int zero = 0, ones = 0, twos = 0;
        Node* curr = head;
    
        // count 0s, 1s, 2s
        while (curr) {
            if (curr->data == 0) zero++;
            else if (curr->data == 1) ones++;
            else twos++;
            curr = curr->next;
        }
    
        // overwrite nodes
        curr = head;
        while (curr) {
            if (zero) {
                curr->data = 0;
                zero--;
            } else if (ones) {
                curr->data = 1;
                ones--;
            } else {
                curr->data = 2;
                twos--;
            }
            curr = curr->next;
        }
    
        return head;
    }
};